plume-common/src/activity_pub/inbox.rs

```diff
@@ -1,48 +1,560 @@
+/// #     type Output = ();
+/// #
+/// #     fn activity(self, _: &(), _actor: User, _id: &str) -> Result<(), ()> {
+/// #         Ok(())
+/// #     }
+/// # }
+/// #
+/// # let mut act = Create::default();
+/// # act.object_props.set_id_string(String::from("https://test.ap/activity")).unwrap();
+/// # let mut person = Person::default();
+/// # person.object_props.set_id_string(String::from("https://test.ap/actor")).unwrap();
+/// # act.create_props.set_actor_object(person).unwrap();
+/// # act.create_props.set_object_object(Note::default()).unwrap();
+/// # let activity_json = serde_json::to_value(act).unwrap();
+/// #
+/// # let conn = ();
+/// #
+/// let result: Result<(), ()> = Inbox::handle(&conn, activity_json)
+///    .with::<User, Announce, Message>()
+///    .with::<User, Create,   Message>()
+///    .done();
+/// ```
+pub enum Inbox<'a, C, E, R> where E: From<InboxError<E>> + Debug {
+    /// The activity has not been handled yet
+    ///
+    /// # Structure
+    ///
+    /// - the context to be passed to each handler.
+    /// - the activity
+    /// - the reason it has not been handled yet
+    NotHandled(&'a C, serde_json::Value, InboxError<E>),
 
-#[derive(Fail, Debug)]
-pub enum InboxError {
-    #[fail(display = "The `type` property is required, but was not present")]
-    NoType,
-    #[fail(display = "Invalid activity type")]
-    InvalidType,
-    #[fail(display = "Couldn't undo activity")]
-    CantUndo,
+    /// A matching handler have been found but failed
+    ///
+    /// The wrapped value is the error returned by the handler
+    Failed(E),
+
+    /// The activity was successfully handled
+    ///
+    /// The wrapped value is the value returned by the handler
+    Handled(R),
 }
 
-pub trait FromActivity<T: Object, C>: Sized {
-    type Error: From<ApError>;
+/// Possible reasons of inbox failure
+#[derive(Debug)]
+pub enum InboxError<E: Debug> {
+    /// None of the registered handlers matched
+    NoMatch,
+
+    /// No ID was provided for the incoming activity, or it was not a string
+    InvalidID,
+
+    /// The activity type matched for at least one handler, but then the actor was
+    /// not of the expected type
+    InvalidActor(Option<E>),
 
-    fn from_activity(conn: &C, obj: T, actor: Id) -> Result<Self, Self::Error>;
+    /// Activity and Actor types matched, but not the Object
+    InvalidObject(Option<E>),
+
+    /// Error while dereferencing the object
+    DerefError,
+}
 
-    fn try_from_activity(conn: &C, act: Create) -> Result<Self, Self::Error> {
-        Self::from_activity(
-            conn,
-            act.create_props.object_object()?,
-            act.create_props.actor_link::<Id>()?,
-        )
+impl<T: Debug> From<InboxError<T>> for () {
+    fn from(_: InboxError<T>) {
+        ()
     }
 }
 
-pub trait Notify<C> {
-    type Error;
+/*
+ Type arguments:
+ - C: Context
+ - E: Error
+ - R: Result
+*/
+impl<'a, C, E, R> Inbox<'a, C, E, R> where E: From<InboxError<E>> + Debug {
```

Comment:
what does the `where` clause here mean? I've never seen `+ Debug` @_@

Reply:
This mean that E must Implement `From<InboxError<E>>` and `Debug` (`Debug` allow ugly display via `println!("{:?}", yourStruct)`) for the functions in this impl block to be available